### ISSUE TITLE
Akka persistence type side effects and at-least-once / at-most-once semantics

### ISSUE THREAD

**Issue by svroonland** [akka/akka#25619]
The akka persistence typed documentation mentions that side effects should be performed after persistent the event, but doesn't mention why. If I'm correct this implies at-least-once semantics for the execution of those side-effects.

I would argue that depending on your application, you might want at-_most_-once semantics, in which case you would want to execute the side effects before persisting the event.

Would be nice to have this in the documentation explicitly, assuming my above thoughts are right of course.

**Comment by patriknw** [comment]
I agree, should be documented. I think we should have a good example showing how to retry side effects until acknowledged, including retry unacknowledged after recovery.

If you do the side effect first you risk that the jvm crashes right after that or that persisting the event fails. Then you end up with a side effect without any event. You rarely want to do this for important side effects.

If you do the side effect after successful persist you risk that the jvm crashes right before the side effect and you end up with an event without having performed the side effect. Typically you want to confirm side effect and then store another event that it has been performed. With that approach you can achieve at-least-once, and effectively once-and-only-once if side effects are idempotent (ok to retry several times).

**Comment by suyambuganesh82** [comment]
@patriknw I was going through the documentation about side effects guarantees and I have the same question about retry until the action is acknowledged. And my google search led here. Is there any example to retry side effects and retry unacknowledged after recovery?

https://doc.akka.io/docs/akka/current/typed/persistence.html#side-effects-ordering-and-guarantees
